### PR TITLE
Reduce recurring D1 reads without KV

### PR DESCRIPTION
**Pull request by Kieren-Foenander**
## Summary

Reduces recurring production D1 reads at their source instead of adding a KV cache. The 24-hour audit showed 819,610 rows read, with the largest repeat costs coming from retention scans and automatic TV preparation retries.

```diff
- run retention every 15 minutes, usually deleting nothing
- build 20 schedule positions and discard 15
- reload and recount the complete Preparation Run per batch
- rediscover canonical show metadata on every retry
- retry every unresolved item every 5 minutes
+ run a complete indexed retention sweep once daily at 03:00 UTC
+ request exactly the 5 schedule positions preparation uses
+ load only the due 5-item batch and count once per round
+ persist canonical identity in the Preparation snapshot
+ progressively retry misses after 5, 15, then 30 minutes
```

## Read flow

```mermaid
flowchart TD
    P[15-minute preparation cron] --> H[Eligible Households only]
    H --> B[Load due five-item batches]
    B --> O{Selection outcome}
    O -->|Downloading| R5[Retry in 5 minutes]
    O -->|Miss: attempts 1-3| R5
    O -->|Miss: attempts 4-7| R15[Retry in 15 minutes]
    O -->|Miss: attempt 8+| R30[Retry in 30 minutes]
    O -->|Ready| Done[Stop reading that item]

    D[Daily 03:00 UTC retention cron] --> C[Walk Households in bounded batches]
    C --> I[Indexed obsolete-state cleanup]
```

## What changed

- Adds migration `0024_reduce_recurring_d1_reads.sql` with indexes for Movie rotation cleanup, stream identity resolution, and video-first canonical episode lookup.
- Stores show IMDb identity, release information, and `next_attempt_at` on Preparation items.
- Keeps active-download polling responsive at five minutes while backing off repeated source misses.
- Avoids full-run item reads and repeated unfinished counts inside every batch.
- Separates automatic preparation and retention into distinct cron routes.
- Walks the complete Household keyspace during daily retention while keeping every SQL batch bounded.
- Documents the measured queries, their purpose, and the expected read reduction.

## Why the retention query existed

Movie rotation rows represent the shuffled lineup for a particular Channel cycle. After the Channel moves to a newer cycle, older rotation rows are disposable because visible playback history is stored separately. The old job checked for those obsolete rows every fifteen minutes and lacked the Household-first access path, so even an empty cleanup was expensive. This change preserves the current cycle and moves the indexed garbage collection to one daily run.

## Verification

- `pnpm test` — 79 Worker tests and 3 component tests pass
- `pnpm typecheck` — passes
- Migration `0024` applies successfully to local D1
- Local `EXPLAIN QUERY PLAN` selects all three new indexes
- `git diff --check` — passes

## Rollout

Apply D1 migration `0024` before deploying the Worker because the new code writes the added Preparation-item columns. After deployment, compare a complete 24-hour D1 Insights window before considering KV.

### CODE REVIEW

File: src/tv-preparation.ts [hunk 1]
```diff
@@ -483,19 +553,31 @@ export class TvSchedulePreparationWorkflow extends WorkflowEntrypoint<TvPreparat
...
+          const result = await step.do(`prepare round ${round + 1} batch ${batch + 1}`, { retries: { limit: 3, delay: "10 seconds", backoff: "exponential" }, timeout: "10 minutes" },
             async () => {
               if (!this.env.CONFIG_SECRET) throw new Error("Configuration secret is unavailable");
               // Keep the decrypted credential inside the step callback so it is never persisted as Workflow output.
               const token = await loadTorBoxCredential(this.env.DB, householdId, this.env.CONFIG_SECRET);
               if (!token) throw new Error("TorBox is not configured");
               return processBatch(this.env, runId, householdId, token, batch);
             });
-          if (complete) return;
+          if (result.stop) return;
```
Comment (Kieren-Foenander) [inline src/tv-preparation.ts]: **[P2] Make the Workflow step graph safe for the nine active production Preparation Runs, and accept cached boolean batch results.**

This `run()` graph is not compatible with instances already in flight:

- New steps are inserted (`load preparation plan`, `finish round N`) before the existing `wait for round N` sleeps.
- `processBatch` used to return a boolean; in-flight steps have that boolean cached. `result.stop` is then `undefined`, so a cached `true` (run already finished or cancelled in that batch) no longer stops the loop.
- The inner loop used to always execute five batches. It now executes `plan.batchCount` batches, so a 5-item run will never replay cached `prepare round N batch 2..5` steps.

Migration `0024` fills `show_imdb_id` / `release_info` but leaves `next_attempt_at` NULL. After deploy, a partially replayed round can mix old batches (NULL `next_attempt_at`) with new batches (5/15/30 minute timestamps). Combined with `MIN()` ignoring NULLs, those due items can wait for a sibling's later wake time.

Cloudflare Workflows match cache by step name, not position; inserting steps in the middle of a live graph is a known break. Suggested fix:

1. Treat a cached boolean as `stop`: `if (result === true || result?.stop) return;`
2. Keep executing five batch steps per round (empty ranges are cheap) so in-flight names still match, **or** terminate `TV_PREPARATION` instances after applying `0024` and document that in the rollout notes.
3. Add a test that a persisted `{ stop: true }` / `true` both end the round.

File: src/tv-preparation.ts [hunk 2]
```diff
@@ -451,21 +497,45 @@ async function processBatch(
...
+    .bind(householdId, runId).first<RunRow>();
+  if (!run || run.status === "cancelled" || run.status === "failed" || run.status === "completed") {
+    return { complete: true, sleepMs: 0 };
+  }
+  const now = new Date();
+  if (Date.parse(run.deadline_at) <= now.getTime()) {
+    await finishRun(db, runId, now, "Preparation window ended");
+    return { complete: true, sleepMs: 0 };
+  }
+  const unfinished = await db.prepare(`SELECT COUNT(*) AS count, MIN(next_attempt_at) AS next_attempt_at
```
Comment (Kieren-Foenander) [inline src/tv-preparation.ts]: **[P2] Treat NULL `next_attempt_at` as due now when computing the Workflow sleep.**

`processBatch` correctly treats `next_attempt_at IS NULL` as due. This aggregate does not: SQLite `MIN()` skips NULLs unless every value is NULL.

Reachable case after this change:

1. Unfinished items still have NULL `next_attempt_at` (new column, migration `0024` does not backfill it, and `createTvPreparationRun` still inserts NULL).
2. At least one sibling in the same Preparation Run already has a future timestamp (a later batch in the round ran the new updater, or a step retried after a partial `processBatch` update).
3. `MIN(next_attempt_at)` returns that future timestamp.
4. The Workflow sleeps 5–30 minutes while the NULL items would have been selected immediately by the due-filter.

That stalls active-download polling as well as source misses, which is the opposite of the intended 5-minute download cadence.

Use a due-now coalescing expression, for example:

```sql
SELECT COUNT(*) AS count,
       MIN(COALESCE(next_attempt_at, ?)) AS next_attempt_at
  FROM tv_preparation_items
 WHERE run_id = ?
   AND status NOT IN ('ready', 'unavailable', 'cancelled')
```

with `now.toISOString()` as the COALESCE value. Please add a test with mixed NULL and future `next_attempt_at` rows that expects a ~1s–immediate wake, not the later sibling's delay.

**Review by Kieren-Foenander — COMMENTED** [review]
## Review

The query/control-flow approach is the right one: snapshot canonical identity, bound the schedule read to five positions, and stop running empty retention every 15 minutes. Two defects should be fixed before merge; both are introduced by this change and are reachable from the new Workflow path.

### Findings

1. **[P2] Make the Workflow step graph safe for the nine active production Preparation Runs, and accept cached boolean batch results** — `src/tv-preparation.ts:575`
2. **[P2] Treat NULL `next_attempt_at` as due now when computing the Workflow sleep** — `src/tv-preparation.ts:525`

### Test gaps

- No test covers `finishRound` sleep calculation, `next_attempt_at` due-filtering, or mixed NULL/timestamp unfinished items.
- Retry backoff is unit-tested in isolation only; the Workflow round loop is untested.
- The `canonical_show_episodes` EXPLAIN assertion was dropped to an index-existence check, so the fallback lookup is not shown to use `canonical_show_episodes_video_idx`.

### Residual risk

Production currently has nine active Preparation Runs. Migration `0024` is necessary before deploy, but it does not make the Workflow step graph backwards-compatible. Terminate or drain those instances as part of rollout.

File: src/tv-preparation.ts (hunk 1 above)
Comment (Kieren-Foenander) [reply src/tv-preparation.ts]: Valid finding; fixed in 7042ddd. The Workflow now executes all five historical batch step names so in-flight instances can replay cached results, while surplus steps in new runs return before any D1 or credential work. Batch outputs accept both the legacy boolean and current `{ stop }` shape. Regression tests cover legacy `true`, current `{ stop: true }`, and the historical step replay path.

File: src/tv-preparation.ts (hunk 2 above)
Comment (Kieren-Foenander) [reply src/tv-preparation.ts]: Valid finding; fixed in 7042ddd. `finishTvPreparationRound` now computes `MIN(COALESCE(next_attempt_at, now))`, so any unfinished NULL item makes the round immediately due instead of inheriting a sibling’s future retry. Added a regression test with mixed NULL and future timestamps that expects the 1-second minimum wake.
